common/objectcachev2.h

```diff
@@ -0,0 +1,257 @@
+
+        Borrow() : _reader(nullptr) {}
+
+        Borrow(ObjectCacheV2* oc, Box* item, std::shared_ptr<V>&& reader)
+            : _oc(oc),
+              _item(item),
+              _reader(std::move(reader)),
+              _recycle(false) {}
+
+        Borrow(Borrow&& rhs) : _reader(nullptr) { *this = std::move(rhs); }
+
+        Borrow& operator=(Borrow&& rhs) {
+            std::swap(_oc, rhs._oc);
+            std::swap(_item, rhs._item);
+            _reader = std::atomic_exchange(&rhs._reader, _reader);
+            std::swap(_reader, rhs._reader);
+            std::swap(_recycle, rhs._recycle);
+            return *this;
+        }
+
+        ~Borrow() {
+            if (_recycle) {
+                _oc->__update(_item, nullptr);
+            }
+        }
+
+        bool recycle() { return _recycle; }
+
+        bool recycle(bool x) { return _recycle = x; }
+
+        V& operator*() const { return *_reader; }
+        V* operator->() const { return &*_reader; }
+        operator bool() const { return (bool)_reader; }
+    };
+
+    template <typename KeyType, typename Ctor>
+    Borrow borrow(KeyType&& key, Ctor&& ctor, uint64_t cooldown = 0UL) {
+        auto item = __find_or_create_item(std::forward<KeyType>(key));
+        auto r = item->reader();
+        while (!r) {
+            if (item->boxlock.try_lock() == 0) {
+                DEFER(item->boxlock.unlock());
+                r = item->reader();
+                if (!r) {
+                    if (photon::sat_add(item->lastcreate, cooldown) <=
+                        photon::now) {
+                        r = __update(item, ctor());
+                    }
+                    return Borrow(this, item, std::move(r));
+                }
+            }
+            photon::thread_yield();
+            r = item->reader();
+        }
+        return Borrow(this, item, std::move(r));
+    }
+
+    template <typename KeyType>
+    Borrow borrow(KeyType&& key) {
+        return borrow(std::forward<KeyType>(key), [&]() { return new V(); });
+    }
+
+    template <typename KeyType, typename Ctor>
+    Borrow update(KeyType&& key, Ctor&& ctor) {
+        auto item = __find_or_create_item(std::forward<KeyType>(key));
+        auto r = __update(item, ctor());
+        return Borrow(this, item, std::move(r));
+    }
+
+    ObjectCacheV2(uint64_t lifespan)
+        : lifespan(lifespan),
+          _timer(1UL * 1000 * 1000, {this, &ObjectCacheV2::__expire}, true,
+                 photon::DEFAULT_STACK_SIZE) {
+        reclaimer = photon::thread_create11([this] {
+            while (true) {
+                auto r = reclaim_queue.recv();
+                if (!r) {
+                    break;
+                } else {
+                    delete r;
+                }
+            }
+        });
+        photon::thread_enable_join(reclaimer);
+    }
+
+    ~ObjectCacheV2() {
+        _timer.stop();
+        if (reclaimer) {
+            reclaim_queue.template send<PhotonPause>(nullptr);
```

Comment:
is it necessary to use the ```template``` keyword?

Reply:
It is necessary.

Calling template member function needs a `template` prefix. The `send<P>` member function has default template argument `ThreadPause`, but here has to call with argument `PhotonPause`, makes here have to be with a `template` prefix.

Reply:
We have following code in RPC's exmaple:
```cpp
    ExampleServer()
        : skeleton(photon::rpc::new_skeleton()),
          server(photon::net::new_tcp_socket_server()) {
        skeleton->register_service<Testrun, Heartbeat, Echo, ReadBuffer,
                                   WriteBuffer>(this);
    }
```
We call the template member function, ```register_service<...>``` without a ```template``` keyword.

Reply:
Yeah but those code in RPC is a part of normal class.

Here is in template define.

ignore that `template` prefix will leads to erros

```
error: expected primary-expression before ‘>’ token
```
in GCC and

```
Use 'template' keyword to treat 'send' as a dependent template name (fix available)clang(missing_dependent_template_keyword)
```

in clang